packages/stateful/actions/actions/AuthzExec.tsx

```diff
@@ -0,0 +1,183 @@
+import type { MsgWithdrawDelegatorReward } from 'cosmjs-types/cosmos/distribution/v1beta1/tx'
+import type {
+  MsgBeginRedelegate,
+  MsgDelegate,
+  MsgUndelegate,
+} from 'cosmjs-types/cosmos/staking/v1beta1/tx'
+import { useCallback, useMemo } from 'react'
+
+import { LockWithKeyEmoji } from '@dao-dao/stateless'
+import {
+  ActionComponent,
+  ActionMaker,
+  CoreActionKey,
+  UseDecodedCosmosMsg,
+  UseDefaults,
+  UseTransformToCosmos,
+} from '@dao-dao/types/actions'
+import {
+  NATIVE_DENOM,
+  makeRawProtobufMsg,
+  makeStargateMessage,
+} from '@dao-dao/utils'
+
+import { AuthzExecComponent as StatelessAuthzComponent } from '../components'
+
+export enum AuthzExecActionTypes {
+  Delegate = '/cosmos.staking.v1beta1.MsgDelegate',
+  Undelegate = '/cosmos.staking.v1beta1.MsgUndelegate',
+  Redelegate = '/cosmos.staking.v1beta1.MsgBeginRedelegate',
+  ClaimRewards = '/cosmos.distribution.v1beta1.MsgWithdrawDelegatorReward',
+  Custom = 'custom',
+}
+
+interface AuthzExecData {
+  authzExecActionType: AuthzExecActionTypes
+  claimRewards: MsgWithdrawDelegatorReward
+  custom: string
+  delegate: MsgDelegate
+  redelegate: MsgBeginRedelegate
+  undelegate: MsgUndelegate
+}
+
+const useDefaults: UseDefaults<AuthzExecData> = () => ({
+  authzExecActionType: AuthzExecActionTypes.Delegate,
+  grantee: '',
+  claimRewards: {
+    delegatorAddress: '',
+    validatorAddress: '',
+  },
+  custom: '[]',
+  delegate: {
+    amount: { denom: NATIVE_DENOM, amount: '0' },
+    delegatorAddress: '',
+    validatorAddress: '',
+  },
+  undelegate: {
+    amount: {
+      denom: NATIVE_DENOM,
+      amount: '0',
+    },
+    delegatorAddress: '',
+    validatorAddress: '',
+  },
+  redelegate: {
+    delegatorAddress: '',
+    validatorSrcAddress: '',
+    validatorDstAddress: '',
+    amount: {
+      denom: NATIVE_DENOM,
+      amount: '0',
+    },
+  },
+})
+
+const Component: ActionComponent = (props) => {
+  return <StatelessAuthzComponent {...props} options={{}} />
+}
+
+const useDecodedCosmosMsg: UseDecodedCosmosMsg<AuthzExecData> = (
+  msg: Record<string, any>
+) =>
+  useMemo(
+    () =>
+      'stargate' in msg && msg.stargate.typeUrl && msg.stargate.value
+        ? {
+            match: true,
+            data: msg as AuthzExecData,
+          }
+        : { match: false },
+    [msg]
+  )
+
```

Comment:
the goal of this hook is to recognize from a CosmosMsg whether or not it is of this action type. right now, all stargate messages will try to render as authz exec actions 🤪

it's probably good enough to add the check that `type_url` is set to `/cosmos.authz.v1beta1.MsgExec` here

Reply:
Oh, lol. Thanks for the context. :joy:

Was trying to figure out what this was for.